File: backend/src/qts/data/historical/csv_index.py
"""Daily byte-offset indexes for ordered historical CSV files."""

from __future__ import annotations

import json
from bisect import bisect_right
from collections.abc import Mapping
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from qts.data.historical.csv_format import parse_historical_ts_event

DEFAULT_TIMESTAMP_COLUMN = "ts_event"
INDEX_VERSION = 1


def write_historical_csv_index(
    csv_path: str | Path,
    *,
    output_path: str | Path | None = None,
    timestamp_column: str = DEFAULT_TIMESTAMP_COLUMN,
) -> Path:
    """Write a daily sidecar index for an ordered historical CSV file."""

    source_path = Path(csv_path)
    target_path = (
        Path(output_path)
        if output_path is not None
        else source_path.with_suffix(source_path.suffix + ".index.json")
    )
    timestamp_column = _required_text(timestamp_column, "timestamp_column")
    buckets: list[dict[str, Any]] = []
    previous_timestamp: datetime | None = None
    current_bucket_date: str | None = None
    row_count = 0

    with source_path.open("rb") as handle:
        header_line = handle.readline()
        if not header_line:
            raise ValueError(f"historical CSV is empty: {source_path}")
        header_offset = handle.tell()
        columns = _columns(header_line)
        if timestamp_column not in columns:
            raise ValueError(f"timestamp column not found: {timestamp_column}")
        timestamp_index = columns.index(timestamp_column)

        while True:
            row_offset = handle.tell()
            line = handle.readline()
            if not line:
                break
            if not line.strip():
                continue
            timestamp = parse_historical_ts_event(_field(line, timestamp_index))
            if previous_timestamp is not None and timestamp < previous_timestamp:
                raise ValueError(
                    "historical CSV timestamps are out of order: "
                    f"{timestamp.isoformat()} before {previous_timestamp.isoformat()}"
                )
            bucket_date = timestamp.date().isoformat()
            if bucket_date != current_bucket_date:
                buckets.append(
                    {
                        "date": bucket_date,
                        "offset": row_offset,
                        "row_index": row_count,
                    }
                )
                current_bucket_date = bucket_date
            previous_timestamp = timestamp
            row_count += 1

    payload = {
        "buckets": buckets,
        "granularity": "day",
        "header_offset": header_offset,
        "row_count": row_count,
        "source_path": str(source_path),
        "timestamp_column": timestamp_column,
        "version": INDEX_VERSION,
    }
    target_path.parent.mkdir(parents=True, exist_ok=True)
    target_path.write_text(json.dumps(payload, sort_keys=True) + "\n", encoding="utf-8")
    return target_path


def indexed_start_offset(
    csv_path: str | Path,
    *,
    start: datetime | None,
    timestamp_column: str = DEFAULT_TIMESTAMP_COLUMN,
    header_offset: int,
) -> int | None:
    """Return the indexed row offset for the UTC day containing ``start``."""

    if start is None:
        return None
    index_path = Path(csv_path).with_suffix(Path(csv_path).suffix + ".index.json")
    if not index_path.exists():
        return None
    payload = _read_payload(index_path)
    if payload.get("version") != INDEX_VERSION:
        return None
    if payload.get("granularity") != "day":
        return None
    if payload.get("timestamp_column") != timestamp_column:
        return None
    if payload.get("header_offset") != header_offset:
        return None
    buckets = payload.get("buckets")
    if not isinstance(buckets, list) or not buckets:
        return None
    start_date = _start_date(start)
    dates: list[str] = []
    offsets: list[int] = []
    for bucket in buckets:
        if not isinstance(bucket, Mapping):
            return None
        date_value = bucket.get("date")
        offset_value = bucket.get("offset")
        if not isinstance(date_value, str):
            return None
        if isinstance(offset_value, bool) or not isinstance(offset_value, int):
            return None
        dates.append(date_value)
        offsets.append(offset_value)
    index = bisect_right(dates, start_date) - 1
    if index < 0:
        return offsets[0]
    return offsets[index]


def _columns(header_line: bytes) -> tuple[str, ...]:
    return tuple(column.strip() for column in header_line.decode("utf-8").strip().split(","))


def _field(line: bytes, index: int) -> str:
    fields = line.decode("utf-8").rstrip("\r\n").split(",")
    if index >= len(fields):
        raise ValueError("historical CSV row has fewer columns than header")
    return fields[index]


def _read_payload(index_path: Path) -> Mapping[str, Any]:
    payload = json.loads(index_path.read_text(encoding="utf-8"))
    if not isinstance(payload, Mapping):
        raise ValueError(f"historical CSV index must be a JSON object: {index_path}")
    return payload


def _required_text(value: str, field_name: str) -> str:
    text = str(value).strip()
    if not text:
        raise ValueError(f"{field_name} is required")
    return text


def _start_date(start: datetime) -> str:
    if start.tzinfo is None or start.utcoffset() is None:
        return start.date().isoformat()
    return start.astimezone(UTC).date().isoformat()


__all__ = [
    "DEFAULT_TIMESTAMP_COLUMN",
    "indexed_start_offset",
    "write_historical_csv_index",
]

File: rust/crates/data/src/csv_index.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};

use chrono::{DateTime, Utc};
use serde_json::json;
use thiserror::Error;

pub const INDEX_VERSION: i64 = 1;
pub const DEFAULT_TIMESTAMP_COLUMN: &str = "ts_event";

pub fn write_historical_csv_index(path: &Path) -> Result<PathBuf, CsvIndexError> {
    let output_path = path.with_extension(format!(
        "{}.index.json",
        path.extension()
            .and_then(|value| value.to_str())
            .unwrap_or_default()
    ));
    let file = File::open(path)?;
    let mut reader = BufReader::new(file);
    let mut byte_offset: u64 = 0;
    let mut header = String::new();
    let header_bytes = reader.read_line(&mut header)?;
    if header_bytes == 0 {
        return Err(CsvIndexError::EmptyCsv(path.to_path_buf()));
    }
    byte_offset += header_bytes as u64;
    let header_offset = byte_offset;
    let columns = header
        .trim_end_matches(['\r', '\n'])
        .split(',')
        .map(str::trim)
        .collect::<Vec<_>>();
    let timestamp_index = columns
        .iter()
        .position(|column| *column == DEFAULT_TIMESTAMP_COLUMN)
        .ok_or(CsvIndexError::MissingTimestampColumn)?;
    let mut buckets = Vec::new();
    let mut previous_timestamp: Option<DateTime<Utc>> = None;
    let mut current_bucket_date: Option<String> = None;
    let mut row_count: u64 = 0;
    loop {
        let row_offset = byte_offset;
        let mut line = String::new();
        let line_bytes = reader.read_line(&mut line)?;
        if line_bytes == 0 {
            break;
        }
        byte_offset += line_bytes as u64;
        if line.trim().is_empty() {
            continue;
        }
        let fields = line
            .trim_end_matches(['\r', '\n'])
            .split(',')
            .collect::<Vec<_>>();
        let timestamp_text = fields
            .get(timestamp_index)
            .ok_or(CsvIndexError::MalformedRow)?;
        let timestamp = parse_timestamp(timestamp_text)?;
        if let Some(previous) = previous_timestamp {
            if timestamp < previous {
                return Err(CsvIndexError::OutOfOrder);
            }
        }
        let bucket_date = timestamp.date_naive().to_string();
        if current_bucket_date.as_deref() != Some(bucket_date.as_str()) {
            buckets.push(json!({
                "date": bucket_date,
                "offset": row_offset,
                "row_index": row_count,
            }));
            current_bucket_date = Some(timestamp.date_naive().to_string());
        }
        previous_timestamp = Some(timestamp);
        row_count += 1;
    }
    let payload = json!({
        "buckets": buckets,
        "granularity": "day",
        "header_offset": header_offset,
        "row_count": row_count,
        "source_path": path.to_string_lossy(),
        "timestamp_column": DEFAULT_TIMESTAMP_COLUMN,
        "version": INDEX_VERSION,
    });
    let mut output = File::create(&output_path)?;
    writeln!(output, "{}", serde_json::to_string(&payload)?)?;
    Ok(output_path)
}

fn parse_timestamp(value: &str) -> Result<DateTime<Utc>, CsvIndexError> {
    Ok(DateTime::parse_from_rfc3339(value)?.with_timezone(&Utc))
}

#[derive(Debug, Error)]
pub enum CsvIndexError {
    #[error("io error")]
    Io(#[from] std::io::Error),
    #[error("json error")]
    Json(#[from] serde_json::Error),
    #[error("timestamp parse error")]
    Timestamp(#[from] chrono::ParseError),
    #[error("historical CSV is empty: {0}")]
    EmptyCsv(PathBuf),
    #[error("timestamp column not found")]
    MissingTimestampColumn,
    #[error("historical CSV row has fewer columns than header")]
    MalformedRow,
    #[error("historical CSV timestamps are out of order")]
    OutOfOrder,
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    #[test]
    fn writes_python_compatible_daily_index() -> Result<(), Box<dyn std::error::Error>> {
        let dir = std::env::temp_dir().join(format!("qts-rs-index-{}", std::process::id()));
        fs::create_dir_all(&dir)?;
        let path = dir.join("gc.csv");
        fs::write(
            &path,
            "ts_event,rtype,publisher_id,instrument_id,open,high,low,close,volume,symbol\n\
             2026-01-02T14:30:00.000000000Z,,,FUTURE.CME.GC.GCM26,1,1,1,1,1,GCM26\n\
             2026-01-03T14:30:00.000000000Z,,,FUTURE.CME.GC.GCM26,1,1,1,1,1,GCM26\n",
        )?;

        let index_path = write_historical_csv_index(&path)?;
        let payload: serde_json::Value = serde_json::from_str(&fs::read_to_string(index_path)?)?;

        assert_eq!(payload["granularity"], "day");
        assert_eq!(payload["row_count"], 2);
        assert_eq!(payload["buckets"][0]["date"], "2026-01-02");
        assert_eq!(payload["buckets"][1]["date"], "2026-01-03");
        Ok(())
    }
}
